source

```diff
@@ -88992,47 +88998,56 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
-      third-party content, especially if it is not currently visible to the user or if resources are
-      constrained. In such cases, the browser could decide to update the rendering for such content
-      infrequently or never.</p>
-     </li>
+     <li id="step-7-3-rendering-opportunities">
+     <p>A <span>browsing context</span> has a <dfn>rendering opportunity</dfn> if the user agent is
+     currently able to present the contents of the <span>browsing context</span> to the user,
+     accounting for hardware refresh rate constraints and user agent throttling for performance
+     reasons, but considering content presented even if it's outside the viewport.</p>
+
+     <p class="note"><span>Browsing context</span> <span data-x="rendering opportunity">rendering
+     opportunities</span> are determined based on hardware constraints such as display refresh
+     rates and other factors such as page performance or whether the page is in the
+     background. Rendering opportunities typically occur at regular intervals. For example, if the
+     browser is attempting to achieve a 60Hz refresh rate, then rendering opportunities occur at a
+     maximum of every 60th of a second (about 16.7ms). If the browser finds that a <span>browsing
+     context</span> is not able to sustain this rate, it might drop to a more sustainable 30
+     rendering opportunities per second for that <span>browsing context</span>, rather than
+     occasionally dropping frames. (This specification does not mandate any particular model for
+     selecting rendering opportunities.)  Similarly, if a <span>browsing context</span> is not
+     visible, the user agent might decide to drop that page to a much slower 4 rendering
+     opportunities per second, or even less.</p>
+
+     <p>If there are <span data-x="browsing context">browsing contexts</span> <var>B</var> that do
+     not have a <span>rendering opportunity</span>, then remove from <var>docs</var> all
+     <code>Document</code> objects whose <span data-x="concept-document-bc">browsing context</span>
+     is in <var>B</var>.</p> </li>
+
+     <li id="step-7-4-unnecessary-rendering"> <p>If there are <span data-x="browsing
+     context">browsing contexts</span> <var>B</var> for which the user agent believes updating the
+     rendering would have no visible effect and which possess no <code>Document</code> objects with
+     a non-empty <span>list of animation frame callbacks</span>, then remove from <var>docs</var>
+     all <code>Document</code> objects whose <span data-x="concept-document-bc">browsing
+     context</span> is in <var>B</var>. Invoke the <span>mark paint timing</span> algorithm for each
+     <code>Document</code> object removed.</p></li>
+
+     <li><p>If there are <span data-x="browsing context">browsing contexts</span> <var>B</var> for
+     which the user agent believes it's preferrable to skip updating the rendering for other
```

Comment:
I'm not sure how specific to be here.

I suspect that for all cases where there is a rendering opportunity and new things to render, but we choose not to rendering anyways, we don't want to mark paint timing, but the only concrete example I've got of when this could happen is for task coalescing.

The key thing we want to avoid is the presence or lack of content to draw impacting whether or not paint timing is marked, which I believe this accomplishes.